cpp/src/decisiontree/batched-levelalgo/kernels/builder_kernels.cuh

```diff
@@ -174,6 +176,24 @@ CUML_KERNEL void excess_sample_with_replacement_kernel(
   for (int i = 0; i < MAX_SAMPLES_PER_THREAD; ++i)
     mask[i] = 0;
 
+  // Specialize CUB collective types for this thread block
+  typedef cub::BlockRadixSort<IdxT, BLOCK_THREADS, MAX_SAMPLES_PER_THREAD> BlockRadixSortT;
+  typedef cub::BlockAdjacentDifference<IdxT, BLOCK_THREADS> BlockAdjacentDifferenceT;
+  typedef cub::BlockScan<IdxT, BLOCK_THREADS> BlockScanT;
+
+  // Shared memory - union allows reuse for different operations
+  __shared__ union TempStorage {
+    typename BlockRadixSortT::TempStorage sort;
+    typename BlockAdjacentDifferenceT::TempStorage diff;
+    typename BlockScanT::TempStorage scan;
+  } temp_storage;
+
+  // Save a random value for offset calculation before sorting destroys the randomness
+  __shared__ IdxT saved_random_value;
+  __shared__ bool random_value_saved;
```

Comment:
Small note: shared memory is usually declared at the top of the kernel so the allocation size is explicit. Variable lifetimes can’t be shortened by the optimizer in this context, so they persist for the entire kernel execution.